### PR TITLE
Add netapp_e_firmware module

### PR DESCRIPTION
**Pull request by ndswartz**
##### SUMMARY
Module performs controller firmware downloads on NetApp E-Series storage systems.
Add unit and integration tests.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
netapp_e_firmware

##### ADDITIONAL INFORMATION

```paste below
ansible 2.9.0.dev0
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/swartzn/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/swartzn/projects/ansible/ansible/lib/ansible
  executable location = /home/swartzn/projects/ansible/ansible/bin/ansible
  python version = 3.6.7 (default, Oct 22 2018, 11:32:17) [GCC 8.2.0]
```

### CODE REVIEW

**Comment by ndswartz** [comment]
shipit

File: lib/ansible/modules/storage/netapp/na_santricity_firmware.py [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = """
+---
+module: na_santricity_firmware
+version_added: "2.9"
+short_description: NetApp E-Series manage firmware bundles and for legacy systems manage embedded firmware, and nvsram file.
```
Comment (lmprice) [inline lib/ansible/modules/storage/netapp/na_santricity_firmware.py]: "NetApp E-Series manage firmware" -- better short description

File: lib/ansible/modules/storage/netapp/na_santricity_firmware.py [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+        required: true
+    wait_for_completion:
+        description:
+            - This flag will cause module to wait for any upgrade actions to complete.
+        type: bool
+        default: false
+    ignore_health_check:
+        description:
+            - This flag will force firmware to be activated in spite of the health check.
+            - Use at you own risk. Certain non-optimal states could result in data loss.
```
Comment (lmprice) [inline lib/ansible/modules/storage/netapp/na_santricity_firmware.py]: your

File: lib/ansible/modules/storage/netapp/na_santricity_firmware.py [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+        type: str
+        required: true
+    wait_for_completion:
+        description:
+            - This flag will cause module to wait for any upgrade actions to complete.
+        type: bool
+        default: false
+    ignore_health_check:
+        description:
+            - This flag will force firmware to be activated in spite of the health check.
```
Comment (lmprice) [inline lib/ansible/modules/storage/netapp/na_santricity_firmware.py]: "regardless of the result of the health check"

File: lib/ansible/modules/storage/netapp/na_santricity_firmware.py [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+            - Use at you own risk. Certain non-optimal states could result in data loss.
+        type: bool
+        default: false
+"""
+EXAMPLES = """
+- name: Ensure correct firmware versions
+  na_santricity_firmware:
+    ssid: "{{ eseries_ssid }}"
```
Comment (lmprice) [inline lib/ansible/modules/storage/netapp/na_santricity_firmware.py]: Can you give some hard-coded examples here? I find that things like the api_url are difficult to discern.

File: lib/ansible/modules/storage/netapp/na_santricity_firmware.py [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+        self.nvsram_name = None
+        self.firmware_name = None
+        self.is_bundle_cache = None
+        self.firmware_version_cache = None
+        self.nvsram_version_cache = None
+        self.upgrade_required = False
+        self.upgrade_in_progress = False
+        self.module_info = dict()
+
+        self.nvsram_name = re.split(r"/", self.nvsram)[-1]
```
Comment (lmprice) [inline lib/ansible/modules/storage/netapp/na_santricity_firmware.py]: I would consider using os.path.basename() here as well. Technically, the path separator is system dependent.

**Review by lmprice — CHANGES_REQUESTED** [review]

**Review by lmprice — APPROVED** [review]
shipit

**Review by thedoubl3j — COMMENTED** [review]
everything looks good to me, only question I have is the module naming. This deviates from previous e-series module naming?

**Comment by ndswartz** [comment]
@thedoubl3j We've decided to be consistent with the netapp_e prefix and make the adjustments when we move to collections. I'll try to make the necessary changes for both na_santricity_firmware and na_santricity_drive_firmware modules by the end of the week.

**Comment by ndswartz** [comment]
@thedoubl3j I've updated the module's name to reflect the netapp_e prefix.

**Review by carchi8py — APPROVED** [review]

**Review by lmprice — APPROVED** [review]
shipit

**Review by carchi8py — APPROVED** [review]

**Review by thedoubl3j — APPROVED** [review]
thanks for adjusting the name back. lgtm. shipit

**Pull request merged**